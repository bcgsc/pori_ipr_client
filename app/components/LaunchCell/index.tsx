import React, { useCallback } from 'react';
import LaunchIcon from '@material-ui/icons/Launch';
import { ICellRendererParams } from '@ag-grid-community/core';
import { useHistory } from 'react-router-dom';

import './index.scss';

type LaunchCellProps = {
  url: (ident: string) => string;
} & ICellRendererParams;

const LaunchCell = (params: LaunchCellProps): JSX.Element => {
  const history = useHistory();

  const handleClick = useCallback((event) => {
    const reportIdent = params.data.reportIdent ?? params.data.ident;

    // event.button is 1 when middle click is pressed
    if (event.ctrlKey || event.metaKey || event.button === 1) {
      window.open(params.url(reportIdent), '_blank');
    } else {
      history.push({ pathname: params.url(reportIdent) });
    }
  }, [history, params]);

  return (
    <LaunchIcon
      className="launch-cell"
      color="action"
      data-testid="launch-cell"
      onClick={handleClick}
      onMouseUp={handleClick}
    />
  );
};

export default LaunchCell;
