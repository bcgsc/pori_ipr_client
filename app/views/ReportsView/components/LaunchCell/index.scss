.launch-cell {
  margin: 5px 0 0;
  cursor: pointer;
}
