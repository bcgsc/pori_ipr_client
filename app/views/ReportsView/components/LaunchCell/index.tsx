import React from 'react';
import LaunchIcon from '@material-ui/icons/Launch';

import './index.scss';

const LaunchCell = (): JSX.Element => (
  <LaunchIcon className="launch-cell" color="action" />
);

export default LaunchCell;
