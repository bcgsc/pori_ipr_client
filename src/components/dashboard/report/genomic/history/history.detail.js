app.controller('controller.dashboard.report.genomic.history.detail',
['_', '$q', '$scope', '$state', '$mdDialog', '$mdToast', 'api.pog', 'api.pogDataHistory', 'pog', 'entry', 'details', 'tags',
(_, $q, scope, $state, $mdDialog, $mdToast, $pog, $history, pog, entry, details, tags) => {

  scope.entry = entry;
  scope.details = details;
  scope.tags = tags;
  scope.tagSelected = entry.tags;
  scope.newEntry = details[entry.new];
  scope.previousEntry = details[entry.previous];
  scope.ignored = ['ident', 'dataVersion', 'createdAt', 'updatedAt', 'deletedAt'];
  scope.tagSearch = {
    selectedItem: null,
    searchText: null
  };

  scope.action = {
    revert: {
      disableButton: false,
      active: false,
      comment: ""
    },
    restore: {
      disableButton: false,
      active: false,
      comment: ""
    }
  };

  scope.test = "bar";

  scope.cancel = () => { $mdDialog.cancel() };

  scope.changeFields = [];

  scope.changed = (f) => { return (scope.changeFields.indexOf(f) > -1) };

  // Auto-complete search filter
  scope.tagSearch.filter = (query) => {

    let deferred = $q.defer();

    $history(pog.POGID).tag.search(query).then(
      (tags) => {
        deferred.resolve(tags);
      },
      (err) => {
        console.log('Unable to search for tags', err);
      }
    );
    return deferred.promise;
  };

  // Create new tag on API
  scope.newTag = (newTag) => {
    // Create new Tag
    $history(pog.POGID).tag.create({tag: newTag.tag}, entry.ident).then(
      (resp) => {
        // Need to find the tag we made, and replace it.
        let found = false;
        _.forEach(scope.tagSelected, (v,k) => {
          if(v.tag === resp.tag) {
            scope.tagSelected[k] = resp;
            found = true;
          }
        });

        if(!found) scope.tagSelected.push(resp);
        $mdToast.show($mdToast.simple().textContent('The tag has been added').position('bottom left'));
      },
      (err) => {
        console.log('Failed to create new tag');
      }
    );
  };

  // Remove tag on API
  scope.removeTag = (removed) => {
    let cachedTag = angular.copy(removed);

    $history(pog.POGID).tag.remove(removed.ident).then(
      (resp) => {
        let toast = $mdToast.simple()
          .textContent('The tag has been removed.')
          .action('Undo')
          .highlightAction(true)
          .highlightClass('md-acceent')
          .hideDelay(4000)
          .position('bottom left');

        $mdToast.show(toast).then(
          (undo) => {
            // Does the user want to undo?
            if(undo === 'ok') {
              // Add it again!
              scope.newTag({tag: cachedTag.tag});
            }
          } // end undo
        )
      },
      (err) => {
        console.log('Failed to create new tag');
      }
    );
  };


  // Revert the opened change
  scope.revert = () => {
    // Hit API
    $history(pog.POGID).revert(entry.ident, scope.action.revert.comment).then(
      (resp) => {
        // Add to history
        $mdDialog.hide({event: 'revert', data: resp});
      },
      (err) => {
        console.log('Failed to revert', err);
        $mdToast.show($mdToast.simple().textContent('Unable to revert the history entry').position('bottom left'));
      }
    );
  };

  // Restore entry
  scope.restore = () => {
    let cached = angular.copy(entry);
    $history(pog.POGID).restore(entry.ident, scope.action.restore.comment).then(
      (result) => {
        // Add to history
        $mdDialog.hide({event: 'restore', data: cached});
      },
      (err) => {
        $mdToast.show($mdToast.simple().textContent('Unable to restore the history event').position('bottom left'));
      }
    )

  };

  // Transform chip for auto complete
  scope.transformChip = (tag) => {
    // If it is an object, it's already a known chip
    if (angular.isObject(tag)) return tag;

    // Otherwise, create a new one
    return { tag: tag, type: 'new' }
  };


  // Search for changes
  let findChanges = (preVal, newVal) => {
    let ignored = ['dataVersion', 'createdAt', 'updatedAt', 'deletedAt'];
    _.forEach(preVal, (v, k) => {
      if(newVal[k] !== v && ignored.indexOf(k) === -1) {
        scope.changeFields.push(k);
      }
    });

  };

  // If it's a change map changes
  if(entry.type == 'change') {
    findChanges(scope.previousEntry, scope.newEntry);
  }


}]);
