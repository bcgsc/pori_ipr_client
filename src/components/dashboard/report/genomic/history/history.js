app.controller('controller.dashboard.report.genomic.history',
['_', '$q', '$scope', '$state', '$mdDialog', '$mdToast', 'api.pog', 'pog', 'api.pogDataHistory', 'history', 'tags', 'exports',
(_, $q, $scope, $state, $mdDialog, $mdToast, $pog, pog, $history, history, tags, exports) => {

  $scope.history = history;

  _.forEach(exports, (ex) => {
    $scope.history.push(ex);
  });

  $scope.history = _.sortBy($scope.history, 'createdAt').reverse();

  console.log($scope.history);

  $scope.tags = tags;

  $scope.detail = ($event, entry) => {

    $history(pog.POGID).detail(entry.ident).then(
      (details) => {
        $mdDialog.show({
          targetEvent: $event,
          templateUrl: 'dashboard/report/genomic/history/history.detail.html',
          clickOutToClose: false,
          locals: {
            entry: entry,
            details: details,
            tags: tags,
            pog: pog
          },
          controller: 'controller.dashboard.report.genomic.history.detail'
        }).then(
          (result) => {

            // Revert!
            console.log('Hidden dialog', event);
            if(result.event === 'revert') {
              $scope.history = _.concat(result.data, $scope.history);
              $mdToast.show($mdToast.simple().textContent('The history event has been reverted'));
            }

            // Removal
            if(result.event === 'restore') {
              // Find and remove the entry from history
              $scope.history = _.filter($scope.history, (h) => { return (h.ident !== result.data.ident) });
              $mdToast.show($mdToast.simple().textContent('The history event has been restored'));
            }
          },
          (err) => {
            console.log('Canceled dialog', err);
          }
        );
      },
      (err) => {
        console.log('Unable to load details');
      }
    );

  }

}]);
