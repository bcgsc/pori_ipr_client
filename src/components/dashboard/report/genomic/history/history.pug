section.dataHistory
    md-subheader.md-primary Data History

    div(layout="column", flex)

        div(layout="column", flex)

            h2.title History Viewer

            div(layout="column", flex).dataBox
                div.box
                    div(layout="row", flex).heading
                        div(flex) Data Section
                        div(flex) Event
                        div(flex=10) Previous
                        div(flex=10) New
                        div(flex) Comment
                        div(flex=10) User
                        div(flex) Tags
                        div(flex=10) Date
                    div(layout="row", flex, ng-repeat="h in history").row
                        div(flex, ng-click="detail($event, h)").clickable.link {{h.model}}
                        div(flex, ng-class="[h.type]").historyType
                            span {{h.type}}
                        div(flex=10).text-center {{h.previous}}
                        div(flex=10).text-center {{h.new || 'none'}}
                        div(flex).text-elipsis {{h.comment || 'none'}}
                        div(flex=10) {{h.user.firstName}} {{h.user.lastName}}
                        div(flex).tags
                            span(ng-repeat="t in h.tags").pill {{t.tag}}
                            span(ng-if="h.tags.length === 0").text-emphasis none
                        div(flex=10)
                            span.clickable {{h.createdAt | amTimeAgo}}
                                md-tooltip {{h.createdAt | amDateFormat: 'DD/MM/YYYY HH:mm:ss'}}