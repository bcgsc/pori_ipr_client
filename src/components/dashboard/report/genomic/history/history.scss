section.dataHistory {

  div.historyType {

    padding: 2px 0;
    text-align: center;
    &.change span{
      background-color: rgb(161, 87, 174);
      color: rgb(255, 255, 255);
      padding: 2px 5px;
      border-radius: 3px;
    }
    &.remove span{
      background-color: rgb(255, 106, 107);
      color: rgb(255, 255, 255);
      padding: 2px 5px;
      border-radius: 3px;
    }

  }

  .link {
    color: rgb(63,81,181);

    &:hover {
      text-decoration: underline;
    }
  }

  .tags {
    padding: 2px 0;
    text-align: center;
    .pill {
      background-color: rgb(63, 81, 181);
      color: rgb(255, 255, 255);
      padding: 2px 6px;
      border-radius: 3px;
      margin-left: 3px;
    }
  }


}


.dataHistoryDetails {

  .diff {
    .previous, .new {
      padding: 8px;
      background-color: rgb(240, 240, 240);
      border: 1px solid rgb(210, 210, 210);
      border-radius: 3px;
      margin: 0 5px;
      overflow-x: auto;

      .changed {
        background-color: rgb(161, 87, 174);
        color: rgb(255,255,255);
      }

      .row {
        div {
          padding: 2px 0 2px 5px;
        }
      }
    }
  }

  h5 {
    margin: 0 0 5px 10px;
    padding: 0;
  }
  pre {
    padding: 8px;
    background-color: rgb(240, 240, 240);
    border: 1px solid rgb(210, 210, 210);
    border-radius: 3px;
    margin: 0 5px 0 10px;
    overflow-x: auto;
  }

}