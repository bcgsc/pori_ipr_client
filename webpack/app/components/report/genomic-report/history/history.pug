section.dataHistory
    md-subheader.md-primary Data History

    div(layout="column", flex)

        div(layout="column", flex)

            h2.title History Viewer

            div(layout="column", flex).dataBox
                div.box
                    div(layout="row", flex).heading
                        div(flex) Data Section
                        div(flex) Event
                        div(flex=10) Previous
                        div(flex=10) New
                        div(flex) Comment
                        div(flex=10) User
                        div(flex) Tags
                        div(flex=10) Date
                    div(layout="row", flex, ng-show="history.length == 0").row
                        div(flex).text-center No History Events
                    div(layout="row", flex, ng-repeat="h in history").row
                        div(flex, ng-if="h.key") CSV Data Export
                        div(flex, ng-if="h.key").historyType.export
                            span export
                        div(flex=20, ng-if="h.key")

                        div(flex, ng-click="detail($event, h)", ng-if="!h.key").clickable.link {{h.model}}
                        div(flex, ng-class="[h.type]", ng-if="!h.key").historyType
                            span {{h.type}}

                        div(flex=10, ng-if="!h.key").text-center {{h.previous}}
                        div(flex=10, ng-if="!h.key").text-center {{h.new || 'none'}}
                        div(flex).text-elipsis {{h.comment || 'none'}}
                        div(flex=10) {{h.user.firstName}} {{h.user.lastName}}
                        div(flex, ng-if="!h.key").tags
                            span(ng-repeat="t in h.tags").pill {{t.tag}}
                            span(ng-if="h.tags.length === 0").text-emphasis none
                        div(flex, ng-if="h.key").tags
                            span.pill.export CSV-Export
                        div(flex=10)
                            span.clickable {{h.createdAt | amTimeAgo}}
                                md-tooltip {{h.createdAt | amDateFormat: 'DD/MM/YYYY HH:mm:ss'}}