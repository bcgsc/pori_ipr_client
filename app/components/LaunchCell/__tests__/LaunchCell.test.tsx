import React from 'react';
import { screen, render, fireEvent } from '@testing-library/react';
import { MemoryRouter, Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';

import LaunchCell from '..';

describe('LaunchCell', () => {
  test('It renders and matches the snapshot', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <LaunchCell />
      </MemoryRouter>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('It renders the icon', async () => {
    render(
      <MemoryRouter>
        <LaunchCell />
      </MemoryRouter>,
    );

    expect(await screen.findByTestId('launch-cell')).toBeInTheDocument();
  });

  test('It calls the url function with args upon click', async () => {
    const mockUrl = jest.fn();
    const mockIdent = '244aabfa-956b0cf9-04388-bbf4';

    render(
      <MemoryRouter>
        <LaunchCell
          url={mockUrl}
          data={{
            reportIdent: mockIdent,
          }}
        />
      </MemoryRouter>,
    );
    fireEvent.click(await screen.findByTestId('launch-cell'));

    expect(mockUrl).toHaveBeenCalledTimes(1);
    expect(mockUrl).toHaveBeenCalledWith(mockIdent);
  });

  test('It calls the url function and history.push upon click', async () => {
    const mockUrl = jest.fn();
    const mockIdent = '244aabfa-956b0cf9-04388-bbf4';
    const history = createMemoryHistory();
    const pushSpy = jest.spyOn(history, 'push');

    render(
      <Router history={history}>
        <LaunchCell
          url={mockUrl}
          data={{
            ident: mockIdent,
          }}
        />
      </Router>,
    );
    fireEvent.click(await screen.findByTestId('launch-cell'));

    expect(pushSpy).toHaveBeenCalledTimes(1);
    expect(pushSpy).toHaveBeenCalledWith({ pathname: mockUrl(mockIdent) });
  });

  test('It opens a new tab with a middle click', async () => {
    const mockUrl = jest.fn();
    const mockIdent = '244aabfa-956b0cf9-04388-bbf4';

    const openSpy = jest.spyOn(window, 'open').mockImplementation(() => undefined);

    render(
      <MemoryRouter>
        <LaunchCell
          url={mockUrl}
          data={{
            reportIdent: mockIdent,
          }}
        />
      </MemoryRouter>,
    );
    fireEvent.click(await screen.findByTestId('launch-cell'), { button: 1 });

    expect(openSpy).toHaveBeenCalledTimes(1);
    expect(openSpy).toHaveBeenCalledWith(mockUrl(mockIdent), '_blank');
  });
});
