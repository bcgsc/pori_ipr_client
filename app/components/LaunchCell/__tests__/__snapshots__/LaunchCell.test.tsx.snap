// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LaunchCell It renders and matches the snapshot 1`] = `
<DocumentFragment>
  <div
    style="height: 100%;"
  >
    <div
      class="ag-root-wrapper ag-layout-normal ag-ltr"
      ref="eRootWrapper"
    >
      
                
                
      <div
        class="ag-root-wrapper-body ag-layout-normal ag-focus-managed"
        ref="rootWrapperBody"
      >
        <div
          class="ag-tab-guard ag-tab-guard-top"
          role="presentation"
          tabindex="0"
        />
        
                    
        <!--AG-GRID-BODY-->
        <div
          aria-colcount="1"
          aria-rowcount="2"
          class="ag-root ag-unselectable ag-layout-normal"
          ref="gridBody"
          role="grid"
          unselectable="on"
        >
          
        
          <!--AG-HEADER-ROOT-->
          <div
            class="ag-header ag-focus-managed ag-pivot-off"
            ref="headerRoot"
            role="presentation"
            style="height: 26px; min-height: 26px;"
            unselectable="on"
          >
            
            
            <div
              class="ag-pinned-left-header ag-hidden"
              ref="ePinnedLeftHeader"
              role="presentation"
              style="width: 0px; max-width: 0px; min-width: 0px;"
            >
              <div
                aria-rowindex="1"
                class="ag-header-row ag-header-row-column"
                role="row"
                style="top: 0px; height: 25px; width: 0px;"
              />
            </div>
            
            
            <div
              class="ag-header-viewport"
              ref="eHeaderViewport"
              role="presentation"
            >
              
                
              <div
                class="ag-header-container"
                ref="eHeaderContainer"
                role="rowgroup"
                style="width: 200px;"
              >
                <div
                  aria-rowindex="1"
                  class="ag-header-row ag-header-row-column"
                  role="row"
                  style="top: 0px; height: 25px; width: 200px;"
                >
                  <div
                    aria-colindex="1"
                    class="ag-header-cell ag-focus-managed"
                    col-id="0"
                    role="columnheader"
                    style="width: 200px; left: 0px;"
                    tabindex="-1"
                    unselectable="on"
                  >
                    
            
                    <div
                      class="ag-header-cell-resize ag-hidden"
                      ref="eResize"
                      role="presentation"
                    />
                    
            
                    <!--AG-CHECKBOX-->
                    <div
                      class="ag-header-select-all ag-labeled ag-label-align-right ag-checkbox ag-input-field ag-hidden"
                      ref="cbSelectAll"
                      role="presentation"
                    >
                      
                
                      <div
                        class="ag-input-field-label ag-label ag-hidden ag-checkbox-label"
                        ref="eLabel"
                        role="presentation"
                      />
                      
                
                      <div
                        class="ag-wrapper ag-input-wrapper ag-checkbox-input-wrapper"
                        ref="eWrapper"
                        role="presentation"
                      >
                        
                    
                        <input
                          aria-label="Press Space to toggle all rows selection (unchecked)"
                          class="ag-input-field-input ag-checkbox-input"
                          id="ag-22-input"
                          ref="eInput"
                          tabindex="-1"
                          type="checkbox"
                        />
                        
                
                      </div>
                      
            
                    </div>
                    
        
                    <div
                      class="ag-cell-label-container"
                      role="presentation"
                    >
                      
            
            
                      <div
                        class="ag-header-cell-label"
                        ref="eLabel"
                        role="presentation"
                        unselectable="on"
                      >
                        
                
                        <span
                          class="ag-header-cell-text"
                          ref="eText"
                          unselectable="on"
                        />
                        
                
                        <span
                          aria-hidden="true"
                          class="ag-header-icon ag-header-label-icon ag-filter-icon ag-hidden"
                          ref="eFilter"
                        >
                          <span
                            class="ag-icon ag-icon-filter"
                            role="presentation"
                            unselectable="on"
                          />
                        </span>
                        
                
                
                
                
            
                      </div>
                      
        
                    </div>
                  </div>
                </div>
              </div>
              
            
            </div>
            
            
            <div
              class="ag-pinned-right-header ag-hidden"
              ref="ePinnedRightHeader"
              role="presentation"
              style="width: 0px; max-width: 0px; min-width: 0px;"
            >
              <div
                aria-rowindex="1"
                class="ag-header-row ag-header-row-column"
                role="row"
                style="top: 0px; height: 25px; width: 0px;"
              />
            </div>
            
        
          </div>
          
        
          <div
            class="ag-floating-top"
            ref="eTop"
            role="presentation"
            style="min-height: 0px; height: 0px; display: none; overflow-y: hidden;"
            unselectable="on"
          >
            
            
            <!--AG-ROW-CONTAINER-->
            <div
              class="ag-pinned-left-floating-top ag-hidden"
              name="topLeft"
              ref="topLeftContainer"
              role="presentation"
              unselectable="on"
            />
            
            
            <!--AG-ROW-CONTAINER-->
            <div
              class="ag-floating-top-viewport"
              name="topCenter"
              ref="topCenterContainer"
              role="presentation"
              unselectable="on"
            >
              
                
              <div
                class="ag-floating-top-container"
                ref="eContainer"
                role="presentation"
                style="width: 200px;"
                unselectable="on"
              />
              
            
            </div>
            
            
            <!--AG-ROW-CONTAINER-->
            <div
              class="ag-pinned-right-floating-top ag-hidden"
              name="topRight"
              ref="topRightContainer"
              role="presentation"
              unselectable="on"
            />
            
            
            <!--AG-ROW-CONTAINER-->
            <div
              class="ag-floating-top-full-width-container"
              name="topFullWidth"
              ref="topFullWidthContainer"
              role="presentation"
              unselectable="on"
            />
            
        
          </div>
          
        
          <div
            class="ag-body-viewport ag-layout-normal ag-row-no-animation"
            ref="eBodyViewport"
            role="presentation"
          >
            
            
            <!--AG-ROW-CONTAINER-->
            <div
              class="ag-pinned-left-cols-container ag-hidden"
              name="left"
              ref="leftContainer"
              role="presentation"
              style="height: 25px;"
              unselectable="on"
            >
              <div
                aria-label="Press SPACE to select this row."
                aria-rowindex="2"
                class="ag-row ag-row-no-focus ag-row-even ag-row-level-0 ag-row-position-absolute ag-row-first ag-row-last"
                comp-id="24"
                role="row"
                row-id="0"
                row-index="0"
                style="height: 25px; transform: translateY(0px); "
              />
            </div>
            
            
            <!--AG-ROW-CONTAINER-->
            <div
              class="ag-center-cols-clipper"
              name="center"
              ref="centerContainer"
              role="presentation"
              style="height: 25px;"
              unselectable="on"
            >
              
                
              <div
                class="ag-center-cols-viewport"
                ref="eViewport"
                role="presentation"
                style="height: 100%;"
              >
                
                    
                <div
                  class="ag-center-cols-container"
                  ref="eContainer"
                  role="rowgroup"
                  style="width: 200px; height: 25px;"
                  unselectable="on"
                >
                  <div
                    aria-label="Press SPACE to select this row."
                    aria-rowindex="2"
                    class="ag-row ag-row-no-focus ag-row-even ag-row-level-0 ag-row-position-absolute ag-row-first ag-row-last"
                    comp-id="25"
                    role="row"
                    row-id="0"
                    row-index="0"
                    style="height: 25px; transform: translateY(0px); "
                  >
                    <div
                      aria-colindex="1"
                      class="ag-cell ag-cell-not-inline-editing ag-cell-auto-height ag-cell-value"
                      col-id="0"
                      comp-id="26"
                      role="gridcell"
                      style="width: 200px; left: 0px;  "
                      tabindex="-1"
                      unselectable="on"
                    />
                  </div>
                </div>
                
                
              </div>
              
            
            </div>
            
            
            <!--AG-ROW-CONTAINER-->
            <div
              class="ag-pinned-right-cols-container ag-hidden"
              name="right"
              ref="rightContainer"
              role="presentation"
              style="height: 25px;"
              unselectable="on"
            >
              <div
                aria-label="Press SPACE to select this row."
                aria-rowindex="2"
                class="ag-row ag-row-no-focus ag-row-even ag-row-level-0 ag-row-position-absolute ag-row-first ag-row-last"
                comp-id="27"
                role="row"
                row-id="0"
                row-index="0"
                style="height: 25px; transform: translateY(0px); "
              />
            </div>
            
            
            <!--AG-ROW-CONTAINER-->
            <div
              class="ag-full-width-container"
              name="fullWidth"
              ref="fullWidthContainer"
              role="presentation"
              style="height: 25px;"
              unselectable="on"
            />
            
        
          </div>
          
        
          <div
            class="ag-floating-bottom"
            ref="eBottom"
            role="presentation"
            style="min-height: 0px; height: 0px; display: none; overflow-y: hidden;"
            unselectable="on"
          >
            
            
            <!--AG-ROW-CONTAINER-->
            <div
              class="ag-pinned-left-floating-bottom ag-hidden"
              name="bottomLeft"
              ref="bottomLeftContainer"
              role="presentation"
              unselectable="on"
            />
            
            
            <!--AG-ROW-CONTAINER-->
            <div
              class="ag-floating-bottom-viewport"
              name="bottomCenter"
              ref="bottomCenterContainer"
              role="presentation"
              unselectable="on"
            >
              
                
              <div
                class="ag-floating-bottom-container"
                ref="eContainer"
                role="presentation"
                style="width: 200px;"
                unselectable="on"
              />
              
            
            </div>
            
            
            <!--AG-ROW-CONTAINER-->
            <div
              class="ag-pinned-right-floating-bottom ag-hidden"
              name="bottomRight"
              ref="bottomRightContainer"
              role="presentation"
              unselectable="on"
            />
            
            
            <!--AG-ROW-CONTAINER-->
            <div
              class="ag-floating-bottom-full-width-container"
              name="bottomFullWidth"
              ref="bottomFullWidthContainer"
              role="presentation"
              unselectable="on"
            />
            
        
          </div>
          
        
          <!--AG-FAKE-HORIZONTAL-SCROLL-->
          <div
            aria-hidden="true"
            class="ag-body-horizontal-scroll"
            style="height: 0px; max-height: 0px; min-height: 0px;"
          >
            
            
            <div
              class="ag-horizontal-left-spacer"
              ref="eLeftSpacer"
              style="width: 0px; max-width: 0px; min-width: 0px;"
            />
            
            
            <div
              class="ag-body-horizontal-scroll-viewport"
              ref="eViewport"
              style="height: 0px; max-height: 0px; min-height: 0px;"
            >
              
                
              <div
                class="ag-body-horizontal-scroll-container"
                ref="eContainer"
                style="height: 0px; max-height: 0px; min-height: 0px; width: 200px;"
              />
              
            
            </div>
            
            
            <div
              class="ag-horizontal-right-spacer"
              ref="eRightSpacer"
              style="width: 0px; max-width: 0px; min-width: 0px;"
            />
            
        
          </div>
          
        
          <!--AG-OVERLAY-WRAPPER-->
          <div
            aria-hidden="true"
            class="ag-overlay ag-hidden"
          >
            
            
            <div
              class="ag-overlay-panel"
            >
              
                
              <div
                class="ag-overlay-wrapper ag-layout-normal"
                ref="eOverlayWrapper"
              />
              
            
            </div>
            
        
          </div>
          
    
        </div>
        
                    
                
        <div
          class="ag-tab-guard ag-tab-guard-bottom"
          role="presentation"
          tabindex="0"
        />
      </div>
      
                
                
      <!--AG-PAGINATION-->
      <div
        aria-describedby="ag-17-start-page ag-17-start-page-number ag-17-of-page ag-17-of-page-number ag-17-first-row ag-17-to ag-17-last-row ag-17-of ag-17-row-count"
        aria-live="polite"
        class="ag-paging-panel ag-unselectable ag-hidden"
        id="ag-17"
      >
        
                
        <span
          aria-hidden="true"
          class="ag-paging-row-summary-panel"
        >
          
                    
          <span
            class="ag-paging-row-summary-panel-number"
            id="ag-17-first-row"
            ref="lbFirstRowOnPage"
          />
          
                    
          <span
            id="ag-17-to"
          >
            to
          </span>
          
                    
          <span
            class="ag-paging-row-summary-panel-number"
            id="ag-17-last-row"
            ref="lbLastRowOnPage"
          />
          
                    
          <span
            id="ag-17-of"
          >
            of
          </span>
          
                    
          <span
            class="ag-paging-row-summary-panel-number"
            id="ag-17-row-count"
            ref="lbRecordCount"
          />
          
                
        </span>
        
                
        <span
          class="ag-paging-page-summary-panel"
          role="presentation"
        >
          
                    
          <div
            aria-label="First Page"
            class="ag-paging-button"
            ref="btFirst"
            role="button"
            tabindex="0"
          >
            <span
              class="ag-icon ag-icon-first"
              role="presentation"
              unselectable="on"
            />
          </div>
          
                    
          <div
            aria-label="Previous Page"
            class="ag-paging-button"
            ref="btPrevious"
            role="button"
            tabindex="0"
          >
            <span
              class="ag-icon ag-icon-previous"
              role="presentation"
              unselectable="on"
            />
          </div>
          
                    
          <span
            aria-hidden="true"
            class="ag-paging-description"
          >
            
                        
            <span
              id="ag-17-start-page"
            >
              Page
            </span>
            
                        
            <span
              class="ag-paging-number"
              id="ag-17-start-page-number"
              ref="lbCurrent"
            />
            
                        
            <span
              id="ag-17-of-page"
            >
              of
            </span>
            
                        
            <span
              class="ag-paging-number"
              id="ag-17-of-page-number"
              ref="lbTotal"
            />
            
                    
          </span>
          
                    
          <div
            aria-label="Next Page"
            class="ag-paging-button"
            ref="btNext"
            role="button"
            tabindex="0"
          >
            <span
              class="ag-icon ag-icon-next"
              role="presentation"
              unselectable="on"
            />
          </div>
          
                    
          <div
            aria-label="Last Page"
            class="ag-paging-button"
            ref="btLast"
            role="button"
            tabindex="0"
          >
            <span
              class="ag-icon ag-icon-last"
              role="presentation"
              unselectable="on"
            />
          </div>
          
                
        </span>
        
            
      </div>
      
                
            
    </div>
  </div>
</DocumentFragment>
`;
